[draytonwiser] Drayton Wiser Binding initial contribution

andrew-schofield opened the pull request:
Add a new binding that supports the [Drayton Wiser](https://wiser.draytoncontrols.co.uk/) smart heating system.

Currently the binding supports auto-discovery, all current devices and most channels.
Some features are still in development, but the binding is usable currently, and has been tested on 4 different heating setups.

There's a discussion thread in the community [here](https://community.openhab.org/t/drayton-wiser-thermostat-binding/35640).

The development of this binding was mentioned in #3126.

jar file for testing:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/org.openhab.binding.draytonwiser/2.5.0-SNAPSHOT/org.openhab.binding.draytonwiser-2.5.0-SNAPSHOT.jar

andrew-schofield commented:
Not sure why the fail state on Jenkins after my most recent commit. The log suggests that the maven JVM died at the end even though the compile completed successfully.

andrew-schofield commented:
I really don't understand the random build system failures I'm seeing. Why are there errors in other plugins, when all I've changed is code in my plugin, and they passed last time?

andrew-schofield commented:
This PR is no longer WIP and is ready for review. I am aware there are multiple commits that are missing the signed-off-by line, but given you're going to request that I squash all the commits before merging anyway, I don't see this as a blocker to review.

andrew-schofield commented:
I've gone through and signed-off all the historic commits that were causing review errors.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+        <state readOnly="true"></state>
+    </channel-type>
+    
+    <channel-type id="hotWaterDemandState-channel">
```
Same type as `heatChannel1DemandState-channel`, please remove.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+        <description>Eco Mode activation state (EcoIQ)</description>
+    </channel-type>
+    
+    <channel-type id="manualModeState-channel">
```
I am wondering whether eco and manual mode can both be set to "on"... if not, how about providing just one `mode` channel` with 2 options?

andrew-schofield replied:
Both can be on, in the state-model of the boiler controller at least. In manual mode I'm assuming it will still use it's eco-mode logic to maintain the manual setpoint.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+        <state readOnly="false" pattern="%.1f h" min="0" max="3" step="0.5"></state>
+    </channel-type>
+    
+    <channel-type id="roomBoostDuration-channel">
```
Same as `hotWaterBoostDuration-channel`

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+    <channel-type id="masterSchedule-channel" advanced="true">
+        <item-type>String</item-type>
+        <label>Room Master Schedule</label>
+        <description>The complete JSON representation of the room schedule</description>
```
Please do not offer channels with plain JSON as input. Channels are supposed to abstract from the technical details. Please remove this one.

andrew-schofield replied:
Yeah, I did wonder about this one. Ideally I want to be able to read and set the schedules that are maintained by the boiler-controller. i.e. I don't want to have my heating schedule controlled by openHAB, rather use it to set "scenes" which apply a particular schedule to my boiler-controller. The only other way I see is to expose the (up to) 52 setpoints for each room for the schedule.

andrew-schofield replied:
Otherwise if openhab were to control the schedule, it would always operate in manual mode, with openhab changing the setpoints.

andrew-schofield replied:
Is there a precedent for doing this in any other bindings already?
It feels like I'd be writing my own DSL for a schedule, just so I can convert it into the Drayton Schedule DSL.

triller-telekom replied:
There is no such thing as a schedule format in other bindings. However this is impossible for the user to configure via a graphical UI. So I would vote to remove this channel.

In the future this functionality could be provided via a `RuleAction` and a `RuleTemplate` of the automation engine, its on our backlog, see: https://github.com/eclipse/smarthome/projects/3#card-8298857

With this you could provide a basic JSON POJO with place holders and provide `ConfigDescription` for the placeholders in a template. A user could then fill out these place holders and the configured action would send the customized POJO to the device.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+        <state readOnly="true" ></state>
+    </channel-type>
+    
+    <channel-type id="signalStrength-channel" advanced="true">
```
This is a perfect candidate to be replaced by a `system-channel`, see https://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html#channels

Please also have a look if you can map some of your other channels to system-channels and remove your definitions afterwards.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+        <state readOnly="true" ></state>
+    </channel-type>
+    
+    <channel-type id="hotWaterOverride-channel">
```
Same as `heatingOverride-channel` with just a different description, please remove one.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+
+    <!-- Channel Types -->
+    <channel-type id="currentTemperature-channel">
+        <item-type>Number</item-type>
```
You can make use of the units of measurement feature here so people can have the value displaye in Fahrenheit for example: Just use `Number:Temperature` here and send `QuantityType`s to the channel. For details, please see: https://www.eclipse.org/smarthome/blog/2018/02/22/units-of-measurement.html

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+    <channel-type id="currentTemperature-channel">
+        <item-type>Number</item-type>
+        <label>Temperature</label>
+        <description>Current temperature in degrees celsius</description>
```
You can remove the unit after using UOM, here.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+    </channel-type>
+    
+    <channel-type id="currentHumidity-channel">
+        <item-type>Number</item-type>
```
See my comment above regarding UOM.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+    </channel-type>
+    
+    <channel-type id="currentSetPoint-channel">
+        <item-type>Number</item-type>
```
See my comment above regarding UOM.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+        <state readOnly="true" ></state>
+    </channel-type>
+    
+    <channel-type id="batteryLevel-channel" advanced="true">
```
See system channels

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+        <state readOnly="true"/>
+    </channel-type>
+    
+    <channel-type id="unifiedBatteryLevel-channel">
```
What is the purpose of this additional channel if you already have a battery-level?

andrew-schofield replied:
This is a good question. The batteryLevel channel exposes the drayton battery levels of "Good, Poor" etc, whereas the unifiedBatteryLevel exposes a percentage so you can use them in sitemaps.
Same with the signalStrength and unifiedSignalStrength.
I exposed them as I initially though someone might want to actually see the "drayton" version of these values, but I'd appreciate your opinion on this.

triller-telekom commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+    </thing-type>
+    
+    <!-- Room Thing Type -->
+    <thing-type id="room">
```
I terms of mapping the drayton devices to openHAB things I am wondering what the value of such a `room` thing is. As far as I understand there has to be a device (TRV or thermostat) inside the room to provide the values.

From what I can tell by a first look is that the room offers additional information about the windows which are certainly useful. The other channels seem to be the ones from the TRV or thermostats inside. Are they somehow aggregated if there are more than one TRV or thermostat inside the room (though more than one thermostat in a room doesn't make much sense)?

If you remove the TRV and thermostat channels from the room, it leaves the window and schedule information. I left a comment for the schedule channel already that we do not like to have such technical, plain JSON, channels.

To sum up all of the above, I would change the room to a `windowContact`, unless the aggregated data are of some value to the user. Also, how do these potentially aggregated values react on user changes to a TRV thing which they can directly manipulate? Wouldn't this become confusing for a user?

What is your opinion on this?

andrew-schofield replied:
The room concept makes sense to me.
The room temperature is an aggregation (well a min({...}) really) of all the sensor values assigned to that room (any combo of TRVs and thermostats).
There are operations that you perform on a room, such as boosting, or setting to manual mode (i.e. if there were multiple TRVs in a room, you wouldn't just boost one of them).
The other thing worth mentioning is that you don't directly assign setpoints etc to the TRVs, you assign them to the room. The only change you can make directly to a TRV thing is to set the locked/unlocked state (i.e. the child lock).

andrew-schofield commented on addons/binding/org.openhab.binding.draytonwiser/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,429 @@
+            <parameter name="REFRESH" type="integer">
+                <label>Refresh interval</label>
+                <description>Interval in seconds between automatic refreshes</description>
+                <required>true</required>
```
Does it apply the default if someone accidentally clears the value here?

triller-telekom replied:
yes

andrew-schofield commented on addons/binding/org.openhab.binding.draytonwiser/OSGI-INF/.gitignore:
```diff
@@ -0,0 +1 @@
+*.xml
```
Done

andrew-schofield commented on addons/binding/org.openhab.binding.draytonwiser/OSGI-INF/DraytonWiserMDNSDiscoveryParticipant.xml:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Done

triller-telekom commented:
I have answered some of you questions, please have a look.

Also please do not add further features to this PR like the smart plug you just added, because this makes the review more difficult. Otherwise if you plan to add more things before someoen reviews them, please change the name of this PR to "WIP" (work in progress).

andrew-schofield commented:
I've sorted out most of the channel names and types etc. I have added UoM to any channels yet, nor have I removed the json channels.

Question: I tried to add humidity as a system channel, however when I tried to use the binding on my instance of 2.3.0 I got an error message about it not being a supported channel-type. Is system.atmospheric-humidity only in >=2.4.0?

triller-telekom commented:
Thank you for sorting out the channel names so far.

To answer your question: Yes the `system.atmospheric-humidity` is very new and only available in 2.4.0-SNAPSHOT of openHAB so far.

andrew-schofield commented:
I've just tried looking at the UoM documentation and some of the bindings that use it, and I'm afraid I'm thoroughly confused.
When I receive some data from the binding, like a temperature update, do I need to return a QuantityType<Temperature>(value, UnitOfReceivedValue)?
When I'm handling a command (like updating a setpoint), how do I convert the command value to the input unit I actually want?

triller-telekom commented:
> When I receive some data from the binding, like a temperature update, do I need to return a QuantityType(value, UnitOfReceivedValue)?

Yes, exactly.

> When I'm handling a command (like updating a setpoint), how do I convert the command value to the input unit I actually want?

That depends on what type the command is and what kind of data your device expects... If you define the channel to only accept `QuantityType`s then you know what value and unit it is. If you also allow `DecimalType`s (Number instead of Number:UOM_Type) then you have to define what you do with that value.

andrew-schofield commented:
OK, I've added UoM to all the channels I think I reasonably can, with the exception of humidity as I haven't set up a test 2.4.0 instance of openhab to make sure it works. I haven't disabled the schedule channel yet, as I'd like to be able to keep a way of updating a schedule, but I need to think about a better way of doing it.
The Jenkins failure appears to be because of a broken modbus binding on master :(

triller-telekom commented:
Thank you for adding the UoM channels. Can't you test the binding within the IDE and control the hardware from there? This way you could make use of the new systemchanneltype and also test it.

Also as mentioned before the schedule channel where pass plain JSON data to the device is certainly somethign that does not belong into the biding in this form, so I would vote for removing it.

However, we can leave it it for now and once @martinvw or @kaikreuzer do a final review they can comment on this one.

Hilbrand commented:
@andrew-schofield Do you need some help migrating this binding to the new build system?

kaikreuzer commented:
FTR, there seems to be a bug in this binding that causes openHAB to lag heavily, see https://github.com/openhab/openhab2-addons/issues/6215#issuecomment-541499214.

Hilbrand commented:
@andrew-schofield the code looks very much complete, but it requires some adaption/reorganization to bring it up-to-date to the latest changes of this repo. If you agree and can test I can make the changes so we can get this binding merged. wdyt?

chrisweetman commented:
really looking forward to this being in openhab, one of the remaining things in my house I havent got on my openhab panel and will find it really useful!

andrew-schofield commented:
If someone can update it to the new build system that would be great. After that it'll need testing for the bug mentioned above before merging I would think.

Hilbrand commented:
I've started making the changes and also some refactoring, which also should fix the mentioned bug. Because I don't have any of these devices it would be great if you can test if I didn't broke all. I'll let you know when I'm ready.

One question I have. If I understand it correctly the secret value needs to be obtained by the user by calling a specific url on the device when in setup mode? If that is indeed the case. Did you consider to make this call from the binding (triggered under certain configuration settings) and automatically set the the secret value?

andrew-schofield commented:
Yes, but the setup mode involves you connecting to the hub's own access point temporarily, there's no automatic way of doing it.

chrisweetman commented:
I went through the process on OH earlier after installing the eclipse iot add on, getting the secret didn't work from a browser, but was simple using the supplied powershell command from windows laptop on the heat hubs own WiFi. The secret is a much longer string than I thought it would be, but I put it into the page on the things page and it auto discovered all 10 trvs, the rooms and thermostats in my house, very impressed... I'm very new to openhab and although technical, have zero programming knowledge, so not done a panel to show temps etc yet, but it's moving in the right direction for me, major props to you Andrew for the work completed and to all others involved.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Andrew Schofield <notifications@github.com>
Sent: Wednesday, January 15, 2020 10:15:23 PM
To: openhab/openhab-addons <openhab-addons@noreply.github.com>
Cc: chrisweetman <chrisweetman@hotmail.co.uk>; Comment <comment@noreply.github.com>
Subject: Re: [openhab/openhab-addons] [draytonwiser] Drayton Wiser Binding initial contribution (#3168)


Yes, but the setup mode involves you connecting to the hub's own access point temporarily, there's no automatic way of doing it.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://eur05.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenhab%2Fopenhab-addons%2Fpull%2F3168%3Femail_source%3Dnotifications%26email_token%3DAII4B3HOKSMNCEAABXUR4N3Q56DHXA5CNFSM4ENPYDZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJCADAI%23issuecomment-574882177&data=02%7C01%7C%7Cda95e35c3aa04333d18208d79a086b2f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637147233252292705&sdata=rH8CuYPz4epEEMvEOTu8tB%2FG097BX%2BRCKDDGOaWro48%3D&reserved=0>, or unsubscribe<https://eur05.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAII4B3FT5KW4QGWF626L6I3Q56DHXANCNFSM4ENPYDZQ&data=02%7C01%7C%7Cda95e35c3aa04333d18208d79a086b2f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637147233252292705&sdata=5dv7rlacpk9MtJecAdD78ebQ%2B61s5cHnjVpzsALoVGk%3D&reserved=0>.

Hilbrand commented:
@andrew-schofield While working on the code I was wondering about [`getExtendedDeviceProperties(0)`](https://github.com/openhab/openhab-addons/blob/bd1dd8bc2a5ac4dd53ab0bb9d666fad80ee3773c/addons/binding/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/handler/ControllerHandler.java#L135). It used `0` as id. Is that because the Hub always has id 0? and/or are these meta data also available for the devices (thing) or only on the Hub?

andrew-schofield commented:
I think I need to explain some terminology and topology of the system in order to explain this.

The physical system is composed of 4 different types of physical device:
HeatHub (the controller attached to the boiler, can override all heating and hot water, set away mode, eco mode etc) `controllerhandler`
Thermostat (room stat, which can control boost state and temperature for a room) `roomstathandler`
TRV (radiator valve which can control boost state and temperature for a room) `trvhandler`
Smart Plug (controls on/off state of power outlet) `smartplughandler`

All the above devices are openhab _things_ and have "extended properties" which expose things like battery level and signal strength etc. The heathub is always device 0.
Thermostats and TRVs can be read by the API, but not set see `rooms` below.

The individual devices above all communicate with the HeatHub over a zigbee network, therefore a "bridge" thing is used which handles the openhab requests, similar to the way Philips Hue works. This is the `heathubhandler`.

There are also some "virtual devices"
Hot Water (part of the controller, can boost hotwater, set setpoints etc) `hotwaterhandler` This isn't available for all models of HeatHub so it is a separate discoverable device
Rooms (collections of room stats, and trvs. Rooms have setpoints and schedules and these delegate control to the stats to regulate the temperature. They don't have an "device" properties.) `roomhandler`

Hilbrand commented:
Ok thanks for the explanation. I think I understand how it works now. I've completed the work I wanted to do and build a new `kar` file (should work with 2.5.x) https://github.com/Hilbrand/openhab-addons/releases/tag/draytonwiser The code can be found in by branch: https://github.com/Hilbrand/openhab-addons/tree/draytonwiser

I've rebuild the commits because it was to difficult to rebase the branch in this pr. What would be preferred to continue?

andrew-schofield commented:
Hi, I tried the binding on openhab 2.5.1 on Windows and unfortunately it didn't work. The bridge never gets out of the initializing state. There are lots of jetty errors in the log which seem to correlate with the binding polling the hub for data.
I'll have to get eclipse installed at some point probably to debug it.

Hilbrand commented:
@andrew-schofield thanks for testing. Can you send me the logs. Maybe best to send it via direct message on the openHAB forum to not polute this thread to much.

dal31b05 commented:
This is a really useful binding but I have reverted to the February 19 version because of issues with Astro and Network bindings when using the latest snapshot. The Feb 19 version seems to be working fine. 

Happy to support testing on my 2.5.1 system if that helps.

dal31b05 commented:
> Hi, I tried the binding on openhab 2.5.1 on Windows and unfortunately it didn't work. The bridge never gets out of the initializing state. 

I also tried this version and the hub never initialized.

jamesmelville commented:
In case it's helpful to others:

I've updated to the latest release from @Hilbrand and it seems to be working fine for me after a bit of clearing out my old set up.

When I updated to the latest snapshot I found that all my Room Thermostats and TRVs were still present, along with the hub, but the controller and rooms were stuck in various statuses "UNINITIALIZED - HANDLER_CONFIGURATION_PENDING" and "UNINITIALIZED - BRIDGE_UNINITIALIZED".

I noticed some data types seemed to have changed in the plugin as battery levels were showing as ISO DateTime stamps. (I had the items defined as String but the definition had changed from String to Number).

I came across this post which hinted at breaking changes to Thing definitions:
https://community.openhab.org/t/z-wave-device-all-become-uninitialized-handler-configuration-pending-after-upgrade/60530

So I deleted all the Things from the Wiser setup from OpenHab and re-added the hub and Things, and everything started working correctly again.

OpenHab 2.5.2-1, Single Channel Wiser without hot water, room stats, trvs and a smart plug.

cpmeister commented:
@andrew-schofield Are you still interested in having this merged for the 2.5.x releases or do you want this to be put off until the the openhab 3 release?

jamesmelville commented:
Hi @Hilbrand, thank you for all your effort updating this. I've installed your updated plugin, and noticed two issues:

1. I've lost my Boiler Controller. It comes up with the status "OFFLINE - COMMUNICATION_ERROR". 
2. Although discovery of the HeatHub still works, Things related to the hub do not appear in the inbox and cannot be found via scanning, they have to be manually added.

Rolling back to @andrew-schofield's 2.5.0 snapshot resolves these issues, but has other errors:
2020-06-16 23:35:54.335 [ERROR] [core.karaf.internal.FeatureInstaller] - Failed installing 'openhab-binding-draytonwiser',

Is there anything I can do to help debug this? I'm about to install Eclipse and have a crack at my first binding

This is on OpenHAB 2.5.5.

Hilbrand commented:
@jamesmelville Based on your feedback I've made some small changes to the binding related to the controller. I don't know if these fix the issue, but it was code that slightly different from the original, (I assumed some values could be null, which could result in the error you got if these values are valid to be null).

If this doesn't fix the problem can you tell if me it does give a detailed error on the communication error? What you can do is run with log level TRACE (`log:set TRACE org.openhab.binding.draytonwiser` in karaf console) and then also run discovery. It might be discovery not working is related to your first mentioned issue. But if not please check the logs. If you run with TRACE it should print the data returned by the device and also detailed information during discovery. Can you send the log to me either via direct message on openHAB forum. You can also install eclipse, checkout my branch and run the binding in eclipse. Then you should be able to see exactly where it goes wrong.

jamesmelville commented:
Thank you! Looking at your code change, it was probably because my device is heating only, and doesn't have a hot water controller. Drayton Wiser controllers can have 1 or 2 heating channels and may or may not control hot water.

As for auto-detection of Things, it did not seem to resolve the issue. After deleting my Controller and Hub / Bridge, the Hub / Bridge appeared in the inbox a little later, but the Controller never appeared after adding it, not in the inbox, which it did in previous version, nor when specifically searching for a Thing.

I had already upped my logging to TRACE but no errors or warnings were logged, just the standard request / responses. I'll send you a DM with the JSON

I'll also attempt to get the binding running locally.
```
2020-06-17 20:58:33.772 [DEBUG] [iser.internal.handler.HeatHubHandler] - Refreshing devices
2020-06-17 20:58:33.776 [DEBUG] [onwiser.internal.api.DraytonWiserApi] - Sending message to heathub: data/domain/
2020-06-17 20:58:33.912 [TRACE] [onwiser.internal.api.DraytonWiserApi] - Reponse (Status:200): //JSON
2020-06-17 20:58:34.200 [DEBUG] [onwiser.internal.api.DraytonWiserApi] - Sending message to heathub: data/network/Station/
2020-06-17 20:58:34.232 [TRACE] [onwiser.internal.api.DraytonWiserApi] - Reponse (Status:200): //JSON
2020-06-17 20:58:34.246 [DEBUG] [iser.internal.handler.HeatHubHandler] - Finished refreshing devices
```

Hilbrand commented:
@jamesmelville I've updates the binding. I think I know why discovery didn't work, as it missed a specific interface. It should be able to detect everything now.

jamesmelville commented:
@Hilbrand thank you, seems to be working perfectly for me now!

Hilbrand commented:
@andrew-schofield I've been working to clean up some small issues and it seems that my updated version of your binding is now in good condition. What I could do is open a new pull request to get that version merged or you can see how you can work with my adapted version and place it in your pull request. If you can let me know how you want to proceed we can work on getting this binding part of the distribution.

andrew-schofield commented:
@Hilbrand Nice work getting the binding operational again. I'll have to admit I haven't had much time to do anything with this recently. As far as I can tell it's working correctly except for the humidity channel on some of the things (I used simple linking in paper UI, and the humidity channel just doesn't enable). I haven't had a chance to look at the log to see if there is an obvious problem.

I can try moving your code over into this PR to get things moving, but if that doesn't work it probably makes sense to just open a new PR from your own fork.

triller-telekom commented:
@Hilbrand I am really sorry, I am too busy for doing another review. I know that i did some initial reviews 2 years ago, but I just can't do that at the moment.

So i would appreciate if someone else can take over. Also i see that my comments have been auto-closed by github, so they should have been incorporated.

Sorry!

Hilbrand commented:
@triller-telekom No problem. I pinged you because you had requested changes, which blocks the process. If you open the "Changes requested" part you'll see 3 dots on the right hand side next to your name. Clicking on them will show a menu were you can dismiss your review, so it won't be blocking anymore. I did make a number changes to this binding so I expect your feedback has been incorporated.

triller-telekom commented:
I could only approve, but if you have done changes and a review, i will trust you :)

Regards,

Stefan

Hilbrand commented:
@triller-telekom I double checked your comments and they seem to be address with exception of the `masterSchedule` channel discussion.

@andrew-schofield  Just to add just before you merged I fixed the humidity channel. It has type `Number:Humidity`, but that doesn't exist. It must be `Number:Dimensionless` (or it can be a system channel: `system.atmospheric-humidity`) I noticed the readme still has the wrong type. I've updated the kar file I posted earlier which contains this fix. To make it work the Things must be recreated.
The PR-openHAB2-Addons builder fails, likely because here only the pom of the binding was updated to `2.5.7-SNAPSHOT`.

andrew-schofield commented:
@Hilbrand I've just done another rebase to bring the base pom up to 2.5.7-snapshot.

That should resolve the PR build.
To be honest, I think at this point the `masterSchedule` should just be removed until a better solution can be found that integrates with OpenHAB.

andrew-schofield commented:
@Hilbrand I've removed the master schedule channel and updated the readme with the required changes (update humidity and remove schedule info).
Assuming everyone is happy, the last thing should be to squash the commits.

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/api/DraytonWiserApiException.java:
```diff
@@ -0,0 +1,28 @@
+ * @author Hilbrand Bouwkamp - Initial contribution
+ */
+@NonNullByDefault
+public class DraytonWiserApiException extends RuntimeException {
```
Why isn't this a checked exception?

andrew-schofield replied:
Because it isn't something that can be recovered from in general, as it is thrown when the heathub is refusing to respond for whatever reason.

andrew-schofield replied:
@Hilbrand Do you have an opinion on this?

Hilbrand replied:
No particular reason. I guess because it's a low level method it means it a lot of methods have to throw the exception if checked, and that sometimes makes code more difficult (to read) when used with streams. But that doesn't seem to be a problem here. So I changed it to a checked exception in the pull request I made.

andrew-schofield replied:
Resolved in cc3fafd

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/discovery/DraytonWiserDiscoveryService.java:
```diff
@@ -0,0 +1,214 @@
+ */
+@NonNullByDefault
+public class DraytonWiserDiscoveryService extends AbstractDiscoveryService
+implements DiscoveryService, ThingHandlerService, DraytonWiserRefreshListener {
```
fix formatting

andrew-schofield replied:
Presumably to all on one line?

cpmeister replied:
the second line should be indented

andrew-schofield replied:
Yeah, I've just run spotless and fixed these 👍

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/HeatHubHandler.java:
```diff
@@ -0,0 +1,168 @@
+            discoveryService.onRefresh(domain);
+        }
+        getThing().getThings().stream() //
+        .filter(thing -> thing.getHandler() instanceof DraytonWiserRefreshListener) //
+        .map(Thing::getHandler) //
+        .map(DraytonWiserRefreshListener.class::cast) //
+        .forEach(listener -> listener.onRefresh(domain));
```
formatting seems wrong here

cpmeister replied:
these lines should be indented

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/HotWaterHandler.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Override
+    protected void handleCommand(final String channelId, final Command command) {
```
Can you handle a refresh as well?

andrew-schofield replied:
I believe this should be handled by the base class

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/RoomHandler.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    protected void handleCommand(final String channelId, final Command command) {
```
same question

andrew-schofield replied:
As above

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/SmartPlugHandler.java:
```diff
@@ -0,0 +1,147 @@
+    protected void handleCommand(final String channelId, final Command command) {
+        switch (channelId) {
+            case CHANNEL_DEVICE_LOCKED:
+                setDeviceLocked(OnOffType.ON.equals(command));
```
Shouldn't you check if the command is an OnOffType first?

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+
+	<channel-type id="heatChannelDemand-channel">
+		<item-type>Number:Dimensionless</item-type>
+		<label>Channel 1 Heat Demand</label>
```
```suggestion
		<label>Heat Demand</label>
```

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,415 @@
+				<description>Network address of the HeatHub</description>
+			</parameter>
+			<parameter name="secret" type="text" required="true">
+				<label>Authorisation Token</label>
+				<description>Auth token required to access the API</description>
```
password context maybe?

andrew-schofield replied:
It's not really a password though, nor an account specific access token. It's a hardcoded value built into the hub.

cpmeister replied:
leave as is then.

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/HotWaterHandler.java:
```diff
@@ -0,0 +1,147 @@
+            }
+        }
+
+        updateState(new ChannelUID(getThing().getUID(), CHANNEL_HOT_WATER_BOOST_DURATION), DecimalType.ZERO);
```
```suggestion
        updateState(CHANNEL_HOT_WATER_BOOST_DURATION, DecimalType.ZERO);
```

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/DraytonWiserThingHandler.java:
```diff
@@ -0,0 +1,196 @@
+    protected void updateState(final String channelId, final Supplier<State> stateFunction) {
+        final State state = api == null || data == null ? UnDefType.UNDEF : stateFunction.get();
+
+        updateState(new ChannelUID(thingUID, channelId), state);
```
```suggestion
        updateState(channelId, state);
```

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/DraytonWiserThingHandler.java:
```diff
@@ -0,0 +1,196 @@
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final ThingUID thingUID;
```
It should be safe to get rid of this field.

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/HotWaterHandler.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    private State getHotWaterDemandState() {
+        return OnOffType.from(getData().hotWater.size() >= 1
+                && "ON".equalsIgnoreCase(getData().hotWater.get(0).getHotWaterRelayState()));
```
`getData().hotWater` should be saved to a local variable.

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/HotWaterHandler.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    private State getManualModeState() {
+        return OnOffType
+                .from(getData().hotWater.size() >= 1 && "MANUAL".equalsIgnoreCase(getData().hotWater.get(0).getMode()));
```
same here

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/HotWaterHandler.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    private State getSetPointState() {
+        return OnOffType.from(getData().hotWater.size() >= 1
+                && "ON".equalsIgnoreCase(getData().hotWater.get(0).getWaterHeatingState()));
```
and here

Hilbrand commented:
@andrew-schofield if you need some help fixing these last comments let me know. It's not needed to squash the commits. We will do this when the pull request is merged.

andrew-schofield commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/api/DraytonWiserApiException.java:
```diff
@@ -0,0 +1,28 @@
+@NonNullByDefault
+public class DraytonWiserApiException extends RuntimeException {
+
+    public DraytonWiserApiException(final String message) {
+        super(message);
+    }
```
Done

andrew-schofield commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/api/DraytonWiserApi.java:
```diff
@@ -0,0 +1,223 @@
+            logger.debug("Interrupted: {}", e.getMessage(), e);
+            Thread.currentThread().interrupt();
+        } catch (final ExecutionException e) {
+            logger.debug("Execution interrupted: {}", e.getMessage(), e);
```
In this instance the effect is the same though. The `ExecutionException` still has the effect of interruption of execution when thrown here.
I'm happy to change it, but to what? Would changing the debug message suffice?

andrew-schofield replied:
Addressed in eaf43b9

Hilbrand commented:
@andrew-schofield You might have seen it, but I made a pull request against your branch with some fixes. https://github.com/andrew-schofield/openhab2-addons/pull/1

andrew-schofield commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/discovery/DraytonWiserDiscoveryService.java:
```diff
@@ -0,0 +1,215 @@
+
+            thingDiscovered(discoveryResult);
+        }
+    }
```
Addressed in eaf43b9

andrew-schofield commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/DraytonWiserThingHandler.java:
```diff
@@ -0,0 +1,196 @@
+            if (api != null && data != null) {
+                handleCommand(channelUID.getId(), command);
+                // update the state after the heathub has had time to react
+                scheduler.schedule(() -> getHeatHubHandler().refresh(), 5, TimeUnit.SECONDS);
```
Addressed in eaf43b9

andrew-schofield commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/DraytonWiserPropertyHelper.java:
```diff
@@ -0,0 +1,47 @@
+        properties.put("Firmware Version", (T) device.getActiveFirmwareVersion());
+        properties.put("Manufacturer", (T) device.getManufacturer());
+        properties.put("Model", (T) device.getModelIdentifier());
+    }
```
Resolved in cc3fafd

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,411 @@
+	</channel-type>
+
+	<channel-type id="boostRemaining-channel">
+		<item-type>Number</item-type>
```
Please change to `Number:Time`

andrew-schofield replied:
Fixed in 8aa384f

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,411 @@
+	</channel-type>
+
+	<channel-type id="boostDuration-channel">
+		<item-type>Number</item-type>
+		<label>Boost Duration</label>
+		<description>Duration to boost hot water for</description>
+		<state pattern="%.1f h" min="0" max="3" step="0.5"/>
```
I would make the same suggestion here, but since the core doesn't properly support handling QuantityType input ranges I think you should leave this as is for now. At the very least though you should specify the expected input unit in the description (in this case hours).

andrew-schofield replied:
Fixed in 8aa384f

cpmeister commented on bundles/org.openhab.binding.draytonwiser/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,411 @@
+		<item-type>Number:Time</item-type>
+		<label>Boost Remaining</label>
+		<description>How long until the boost deactivates</description>
+		<state readOnly="true" pattern="%.0f min"/>
```
```suggestion
		<state readOnly="true" pattern="%.0f %unit%"/>
```

Hilbrand replied:
I didn't use that because my assumption is that using unit will show it in the unit set in code, which is in seconds, because in code that is the most logical. However the preferred display unit is minutes and is forced here by setting the unit itself.

andrew-schofield replied:
I've "resolved" that by making the boost remaining state return the value in `SmartHomeUnits.MINUTE` since the value isn't used by anything internally for any logic.

cpmeister commented on bundles/org.openhab.binding.draytonwiser/README.md:
```diff
@@ -0,0 +1,248 @@
+| `hotWaterOverride`       | Switch       | State of the hot water override button on the controller |
+| `hotWaterDemandState`    | Switch       | Is hot water calling the boiler for heat                 |
+| `hotWaterBoosted`        | Switch       | Is hot water currently being boosted                     |
+| `hotWaterBoostRemaining` | Number       | How long until the boost deactivates in minutes          |
```
```suggestion
| `hotWaterBoostRemaining` | Number:Time  | How long until the boost deactivates in minutes          |
```

Please update the other BoostRemaning channels in the readme.

Hilbrand commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/HotWaterHandler.java:
```diff
@@ -132,7 +132,11 @@ private State getBoostRemainingState() {
 
             if (overrideTimeout != null && !"NONE".equalsIgnoreCase(firstChannel.getOverrideType())) {
-                return new QuantityType<Time>(overrideTimeout - (System.currentTimeMillis() / 1000L), SmartHomeUnits.SECOND).toUnit(SmartHomeUnits.MINUTE);
+                final QuantityType<Time> duration = new QuantityType<Time>(overrideTimeout - (System.currentTimeMillis() / 1000L), SmartHomeUnits.SECOND).toUnit(SmartHomeUnits.MINUTE);
```
This was a bad idea. I see you now need to do the extra null check just to not get a compile error :sob:, while it doesn't add any value here. Besides the incorrect styling, I think you should just use seconds as unit and not convert it to minutes here. It would even be better to just divide by 60. But that misses the whole point of using smart units.

fwolter commented on bundles/org.openhab.binding.draytonwiser/README.md:
```diff
@@ -0,0 +1,248 @@
+| Channel                      | Item Type            | Description                                              |
+|------------------------------|----------------------|----------------------------------------------------------|
+| `heatingOverride`            | Switch               | State of the heating override button on the controller   |
+| `heatChannel1Demand`         | Number:Dimensionless | Current demand level of heating channel 1                |
```
Is this percent?

Hilbrand replied:
Yes this is generaly used for readonly percentage channels.

fwolter commented on bundles/org.openhab.binding.draytonwiser/README.md:
```diff
@@ -0,0 +1,248 @@
+| `heatRequest`        | Switch               | Is the room actively requesting heat from the controller                     |
+| `roomBoosted`        | Switch               | Is the room currently being boosted                                          |
+| `roomBoostRemaining` | Number:Time          | How long until the boost deactivates in minutes                              |
+| `windowState`        | Switch               | Is the window open or closed?                                                |
```
If I see correctly you used `Contact` in the code.

andrew-schofield replied:
Yes, I'll fix that in the Readme

andrew-schofield replied:
Fixed in 9a6a1e1

fwolter commented on bundles/org.openhab.binding.draytonwiser/README.md:
```diff
@@ -0,0 +1,248 @@
+|----------------------|----------------------|------------------------------------------------------------------------------|
+| `currentTemperature` | Number:Temperature   | Currently reported temperature                                               |
+| `currentHumidity`    | Number:Dimensionless | Currently reported humidity (if there is a room stat configured in this room |
+| `currentDemand`      | Number:Dimensionless | Current heat demand percentage of the room                                   |
```
You could use `PercentType`.

Hilbrand replied:
`PercentType` is a dimmer. If the channel is readonly it's generally prefered to not use a dimmer, but use the `Number:Dimensionless`

fwolter replied:
Oops, mixed that up, sorry.

fwolter commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/api/DraytonWiserApi.java:
```diff
@@ -0,0 +1,226 @@
+        } catch (final TimeoutException e) {
+            failCount++;
+            if (failCount > 2) {
+                logger.debug("Heathub didn't repond in time: {}", e.getMessage(), e);
```
It seems a bit unreasonable to log the stack trace when the network fails.
```suggestion
                logger.debug("Heathub didn't repond in time: {}", e.getMessage());
```

andrew-schofield replied:
Resolved in af8d624

fwolter commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/api/DraytonWiserApi.java:
```diff
@@ -0,0 +1,226 @@
+                throw new DraytonWiserApiException("Heathub didn't repond in time", e);
+            }
+        } catch (final InterruptedException e) {
+            logger.debug("Interrupted: {}", e.getMessage(), e);
```
Log stack trace, see above.

andrew-schofield replied:
Resolved in af8d624

fwolter commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/discovery/DraytonWiserDiscoveryService.java:
```diff
@@ -0,0 +1,201 @@
+            final Map<String, Object> properties = new HashMap<>();
+
+            DraytonWiserPropertyHelper.setControllerProperties(device, properties);
+            final DiscoveryResult discoveryResult = DiscoveryResultBuilder
+                    .create(new ThingUID(THING_TYPE_CONTROLLER, bridgeUID, "controller")).withProperties(properties)
+                    .withBridge(bridgeUID).withLabel("Controller").build();
```
Can you add a representation property? Same for below.

Hilbrand replied:
The controller doesn't have any properties, which might not be good here. I'm not familiar with this device, but it does get the DeviceDTO, which does have a serial number value. So if that makes sense it could be applied here? @andrew-schofield does the controller have a serial number?

andrew-schofield replied:
No :(

andrew-schofield replied:
Resolved in af8d624

fwolter commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/discovery/DraytonWiserMDNSDiscoveryParticipant.java:
```diff
@@ -0,0 +1,90 @@
+                            DraytonWiserBindingConstants.DEFAULT_REFRESH_SECONDS);
+                }
+
+                return DiscoveryResultBuilder.create(uid).withProperties(properties)
+                        .withRepresentationProperty(uid.getId()).withLabel("Heat Hub - " + service.getName()).build();
```
The representation property must be the name of the property.
```suggestion
                return DiscoveryResultBuilder.create(uid).withProperties(properties)
                        .withRepresentationProperty(DraytonWiserBindingConstants.ADDRESS).withLabel("Heat Hub - " + service.getName()).build();
```

Hilbrand replied:
I see I haven't done the representation property correctly. In the thing xml it uses `serialNumber` as representation property. That value is set in the handler. So it should be serialNumber here. That should also be set as representation property of all things having a serial number. And also in the thing xml it should be added. The other representation property is room name and needs to be added to both code and thing xml

andrew-schofield replied:
Is this an issue with the refactor that you did? If so I'll leave you to make the necessary changes.

Hilbrand replied:
yes probably :smile: I'll fix it. Do you know if the controller has a serial number or any identification?

andrew-schofield replied:
Sadly not, no. This is the json representation of the controller:
`{
            "id": 0,
            "NodeId": 0,
            "ProductType": "Controller",
            "ProductIdentifier": "Controller",
            "ActiveFirmwareVersion": "2.54.0",
            "ModelIdentifier": "WT714R1S0902",
            "DeviceLockEnabled": false,
            "DisplayedSignalStrength": "VeryGood",
            "ReceptionOfController": {
                "Rssi": -61
            },
            "ReceptionOfDevice": {
                "Rssi": 0,
                "Lqi": 255
            }
        }`

Unfortunately if you happen to have 2 heating systems, the json response would look identical.

andrew-schofield replied:
Resolved in af8d624

fwolter commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/DraytonWiserThingHandler.java:
```diff
@@ -0,0 +1,222 @@
+        final HeatHubHandler bridgeHandler = getHeatHubHandler();
+
+        if (bridgeHandler == null) {
+            api = null;
```
Better set the Thing status to OFFLINE with an appropriate error message. Otherwise the Thing will stick with INITIALIZING.

Hilbrand replied:
The method `getHeatHubHandler()` already sets it to offline in this case.

fwolter commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/handler/RoomHandler.java:
```diff
@@ -0,0 +1,194 @@
+        switch (channelId) {
+            case CHANNEL_CURRENT_SETPOINT:
+                if (command instanceof QuantityType) {
+                    setSetPoint((QuantityType<Temperature>) command);
```
You could use the wildcard here and in `setSetPoint()` to eliminate the compiler warning.
```suggestion
                    setSetPoint((QuantityType<?>) command);
```

andrew-schofield replied:
Resolved in af8d624

fwolter commented on bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/discovery/DraytonWiserDiscoveryService.java:
```diff
@@ -0,0 +1,201 @@
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerService;
+import org.eclipse.smarthome.core.thing.binding.ThingTypeProvider;
+import org.eclipse.smarthome.core.thing.xml.internal.ThingTypeXmlProvider;
```
These imports are never used.

andrew-schofield replied:
Fixed in 4da5349

andrew-schofield commented:
> LGTM
> 
> There are two approvals from maintainers, now.  But @Hilbrand contributed to this PR. How do we handle this?

Yeah, I would be happier if @cpmeister also approved it if we need more than one approval.

Hilbrand commented:
I agree that my approval is not completely independent 😉 It's just @cpmeister did approve and then left 2 small comments, which have been addressed imoh. And because @cpmeister doesn't seem to be available at the moment (his last activity was july 15th) I don't think that should be blocking this being merged (given he already approved it albeit with 2 additional comments. So if you agree with me @fwolter I would suggest we merge this binding.

kaikreuzer commented:
I agree with @Hilbrand and this PR has anyhow a looong history with a lot of reviews, so that it is really time to have it merged. Big thanks to everyone involved here!
